Calculating the Pilot Residual Lifetime for analysis running jobs.

todor-ivanov opened the pull request.

todor-ivanov committed:
Calculating the Pilot Residual Lifetime for analysis running jobs.

removing few fields from the running jobs

add exception for PilotRestLifeTimeMins

Adding negative default value for PilotRestLifeTimeMins

Adding default value for HasBeenTimingTuned

Handling the KeyError exception explicitly

Implementing the Benjamin's suggestion.

Edit only the result dictionary for PilotRestLifeTimeMins

stiegerb commented on src/htcondor_es/convert_to_json.py:
```diff
@@ -633,6 +639,9 @@ def convert_to_json(ad, cms=True, return_dict=False, reduce_data=False):
     result["WallClockHr"] = ad.get("RemoteWallClockTime", 0)/3600.
 
+    if analysis and ad.get("JobStatus") == 2:
+        result["PilotRestLifeTimeMins"] = int((ad['MATCH_GLIDEIN_ToDie'] - ad['EnteredCurrentStatus'])/60)
```
@todor-ivanov We should make sure this doesn't fail, so either put it inside a try/except clause or put a default value by using `get` instead of the direct lookup. I'm not sure what's better, it depends on how you want to handle the default values.

stiegerb commented on src/htcondor_es/convert_to_json.py:
```diff
@@ -472,6 +477,7 @@
   "DESIRED_SITES_Orig",
   "EstimatedWallTimeMins",
   "EstimatedWallTimeJobCount",
+  "PilotRestLifeTimeMins",
```
@todor-ivanov can you add it also to the `int_vals` list?

stiegerb reviewed (changes requested):
Let's put a default value for the new field and catch cases where the two ingredients aren't filled.

todor-ivanov commented:
@stiegerb I added a check if the job has ever matched a pilot and an exception which will fill the pilot's maximum lifetime as a negative value, which should be an indication of either a missing 'MATCH_GLIDEIN_ToDie' classadd attribute or of some dimensionality problems. 
The default is set to -1. If 'PilotRestLifeTimeMins' is calculated correctly it should never gain a negative value. So it has two different markers for this two situations now (a wrong  calculation or not calculated at all)

todor-ivanov commented:
@stiegerb - I also added a default value for 'HasBeenTimingTuned'. Hope the way I try to do it is a valid one and it does not break anything.

stiegerb commented:
Sorry for being picky, but I find the logic a bit confusing. And I'd prefer to explicitly except the `KeyError`'s there. Would this do the same thing but possibly be easier to read?

```python
result["PilotRestLifeTimeMins"] = -1
if analysis and ad["JobStatus"] == 2 and 'LastMatchTime' in ad:
    try:
        ad["PilotRestLifeTimeMins"] = int((ad['MATCH_GLIDEIN_ToDie'] -
                                           ad['EnteredCurrentStatus'])/60)
    except KeyError:
        ad["PilotRestLifeTimeMins"] = -72*60

result['HasBeenTimingTuned'] = ad.get('HasBeenTimingTuned', False)
```

todor-ivanov commented:
Yes, it seems that  the change you propose will make things more readable. I will make the PR now.

stiegerb commented:
Ok, then if you also add `HasBeenTimingTuned` to the bool_val list and squash all the commits into one I'll merge it.

Thanks

todor-ivanov commented:
'HasBeenTimingTuned' is already in the bool_val list, I think.

todor-ivanov commented:
@stiegerb commits are squashed into a single one now.

stiegerb commented:
@todor-ivanov Thanks, but I still see the more confusing logic from before. Am I missing something?

todor-ivanov commented:
@stiegerb Sorry, it was me but not you the one who was missing a thing here. I just added the change and squashed again. I preserved only this line:
result["PilotRestLifeTimeMins"] = ad.get("PilotRestLifeTimeMins", -1)
Just to be 100% sure that  at the end we have the value set  in the 'result' dictionary not in the 'ad' only.

stiegerb commented:
Ok, I was a bit confused by your comment, but then I read my proposed code snipped again and noticed that I was doing something silly (changing the ad instead of the result). Here's what I want:

```python
result["PilotRestLifeTimeMins"] = -1
if analysis and ad.get("JobStatus") == 2 and 'LastMatchTime' in ad:
    try:
        result["PilotRestLifeTimeMins"] = int((ad['MATCH_GLIDEIN_ToDie'] -
        	                               ad['EnteredCurrentStatus'])/60)
    except KeyError:
        result["PilotRestLifeTimeMins"] = -72*60
```

I.e. setting the default value in `result` first, and then never even touching the `ad`.

Sorry for the confusion.

stiegerb commented:
Perfect, thanks. Can you squash? Then I'll merge and update.

todor-ivanov commented:
The PR should be in the proper format now.

todor-ivanov commented:
Thanks!
